### PR TITLE
CLI migration to GraphQL endpoint

### PR DESCRIPTION
**Pull request by dwsutherland**
As a prerequisite to having the CLI interact with the UIS GraphQL endpoint via the HUB, this PR:
- [x] Migrates the CLI to the one GraphQL endpoint.
- [x] Removes the unused endpoints.
- [x] Removes the Cylc7 data generation & structures.

Addressed elsewhere:
- [x] `cylc_scan.py` dealt with in #3724 

Commands scripts to do:
- [x] `cylc_broadcast.py`
- [x] `cylc_checkpoint.py`
- [x] `cylc_dump.py`
- [x] `cylc_ext_trigger.py`
- [x] `cylc_get_suite_version.py`
- [x] `cylc_hold.py`
- [x] `cylc_kill.py`
- [x] `cylc_message.py`/`task_message.py`
- [x] `cylc_nudge.py`
- [x] `cylc_ping.py`
- [x] `cylc_poll.py`
- [x] `cylc_release.py`
- [x] `cylc_reload.py`
- [x] `cylc_remove.py`
- [x] `cylc_set_outputs.py`
- [x] `cylc_set_verbosity.py`
- [x] `cylc_show.py`
- [x] `cylc_stop.py`
- [x] `cylc_trigger.py`

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Already covered by existing tests.
- [x] No change log entry required (invisible to users).
- [x] No documentation update required.
- [x] No dependency changes.

### CODE REVIEW

**Comment by dwsutherland** [comment]
Slowly making my way through.. 

Broadcast deltas are somewhat event driven (create a `workflow.broadcast` delta on set/clear of broadcasts), this could be the template for fixing #3746

**Comment by dwsutherland** [comment]
#3724 Needs to go in first

File: cylc/flow/scheduler.py [hunk 1]
```diff
@@ -364,7 +360,6 @@ async def initialise(self):
             self.suite, context=self.zmq_context, barrier=self.barrier)
 
         self.proc_pool = SubProcPool()
-        self.state_summary_mgr = StateSummaryMgr()
```
Comment (oliver-sanders) [inline cylc/flow/scheduler.py:367]: Woohoo!

File: cylc/flow/scheduler.py [hunk 2]
```diff
@@ -882,100 +844,22 @@ def info_get_graph_raw(self, cto, ctn, group_nodes=None,
...
-
     def command_stop(
             self,
             mode=None,
             cycle_point=None,
             # NOTE clock_time YYYY/MM/DD-HH:mm back-compat removed
             clock_time=None,
-            task=None
+            task=None,
+            kill_active_tasks=False,
```
Comment (oliver-sanders) [inline cylc/flow/scheduler.py:854]: `kill_active_tasks` is covered by `StopMode.KILL`.

File: cylc/flow/scripts/cylc_get_suite_version.py [hunk 3]
```diff
@@ -32,8 +36,30 @@ def main():
     (options, args) = parser.parse_args()
     suite = args[0]
 
-    pclient = SuiteRuntimeClient(suite, timeout=options.comms_timeout)
-    print(pclient('get_cylc_version'))
+    try:
+        pclient = SuiteRuntimeClient(suite, timeout=options.comms_timeout)
+    except Exception as exc:
+        print(cparse(f'<red>{exc}</red>'))
+        sys.exit(1)
```
Comment (oliver-sanders) [inline cylc/flow/scripts/cylc_get_suite_version.py:43]: OOh, looks like this script isn't wrapped with `cylc.flow.terminal.cli_function` for some reason, that will take care of exception formatting.

File: cylc/flow/scripts/cylc_ping.py [hunk 4]
```diff
@@ -45,22 +47,59 @@ def get_option_parser():
...
+            query ($wFlows: [ID]) {
+              workflows(ids: $wFlows) {
+                id
+                name
+                port
+                pubPort
+              }
+            }
+        ''',
+        'variables': {'wFlows': [suite]}
```
Comment (oliver-sanders) [inline cylc/flow/scripts/cylc_ping.py:64]: Is this using a running workflow to list other running workflows?

Or is the intention to use the UIS?

**Review by oliver-sanders — COMMENTED** [review]
Looks like you had some fun with `cylc broadcast`!

File: cylc/flow/scheduler.py (hunk 1 above)
Comment (dwsutherland) [reply cylc/flow/scheduler.py:367]: There might be some `TaskProxy.summary` to get rid of too.. But I'll leave that for when the data store moves to event driven updates.

File: cylc/flow/scheduler.py (hunk 2 above)
Comment (dwsutherland) [reply cylc/flow/scheduler.py:854]: Oh, right! (need to put another mode in the input enum then).

File: cylc/flow/scripts/cylc_ping.py (hunk 4 above)
Comment (dwsutherland) [reply cylc/flow/scripts/cylc_ping.py:64]: Artifact of UIS use, if the data-store had more than one workflow (every data-store is structured `self.data[w_id]`) or the resolvers used scan to send the query off to other workflows (without causing a network storm! 🌩️ ) we could..

File: cylc/flow/scheduler.py (hunk 2 above)
Comment (dwsutherland) [reply cylc/flow/scheduler.py]: There existed no `StopMode.KILL` so I created the new enum for it (`StopMode.REQUEST_KILL`) and removed `kill_active_tasks`...

File: cylc/flow/scripts/cylc_get_suite_version.py (hunk 3 above)
Comment (dwsutherland) [reply cylc/flow/scripts/cylc_get_suite_version.py]: Done.

File: cylc/flow/scripts/cylc_ping.py (hunk 4 above)
Comment (dwsutherland) [reply cylc/flow/scripts/cylc_ping.py:64]: Also, yes (if that's what you meant, sorry!); once the CLI goes through the UIS this will kick in to target multiple workflows.

**Comment by dwsutherland** [comment]
> No change log entry required (invisible to users).

The only change I can think of that would effect users, is the CLI JSON format is different for a few commands (i.e. `cylc show` )..
The other one; `cylc show --list-prerequisites` only shows prerequisites for live tasks, which may be rectified post n-window (TODO in code)..

File: cylc/flow/scripts/cylc_get_suite_version.py [hunk 5]
```diff
@@ -20,20 +20,42 @@
...
+              }
+            }
+        ''',
+        'variables': {'wFlows': [suite]}
+    }
+
+    result = pclient('graphql', query_kwargs)
+
+    for workflow in result['workflows']:
+        print(workflow['cylcVersion'])
```
Comment (kinow) [inline cylc/flow/scripts/cylc_get_suite_version.py:58]: I think it will always be a single workflow, but harmless to use a loop to unpack the first result I think

**Review by kinow — APPROVED** [review]
Tested with the workflow `five`, found no errors using the few commands I am familiar with (run, nudge, poll, broadcast, hold, release, cat-log, dump).

Had a look at the code, left a few comments, but no blockers for me :+1: amazing work @dwsutherland !

Tried to test the UI, but there's some hiccup with that `MSG_TIMEOUT`, causing UIS -> WFS to fail to spawn in jupyterhub.

Thanks!
B

File: cylc/flow/task_message.py [hunk 6]
```diff
@@ -46,6 +46,24 @@
...
+  message (
+    workflows: $wFlows,
+    taskJob: $taskJob,
+    eventTime: $eventTime,
+    messages: $messages
+  ) {
+    result
+  }
+}
+'''
```
[the remark this answers is not in the document]
Comment (dwsutherland) [reply cylc/flow/task_message.py:65]: > Would it be useful here too? Looks like Graphene provides objects to create mutations too: https://docs.graphene-python.org/en/latest/types/mutations/

Those are schema definition objects.

File: cylc/flow/scripts/cylc_get_suite_version.py (hunk 5 above)
Comment (dwsutherland) [reply cylc/flow/scripts/cylc_get_suite_version.py:58]: UIS API/client can/will peddle to multiple workflows.

File: cylc/flow/scripts/cylc_show.py [hunk 7]
```diff
@@ -57,77 +137,136 @@ def get_option_parser():
...
                             print('  o  %s ... %s' % (key, value))
-            for bad_item in bad_items:
-                ansiprint(
-                    f"<red>No matching tasks found: {bad_item}\n",
-                    file=sys.stderr)
-            if bad_items:
-                sys.exit(1)
+        if not results['taskProxies']:
+            ansiprint(
+                f"<red>No matching tasks found: {task_ids}",
```
[the remark this answers is not in the document]
Comment (dwsutherland) [reply cylc/flow/scripts/cylc_show.py:267]: That code was already there (see above).. (I suppose it just means "do the whole line")

File: cylc/flow/task_message.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dwsutherland) [reply cylc/flow/task_message.py:65]: Could be an idea to collect the strings into one place, but there's no (or almost no) duplication because each is tailored to the respective command (and I can't think of where else I'd use them).

[the remark this answers is not in the document]
Comment (kinow) [reply cylc/flow/task_message.py:65]: I wasn't sure if there were queries or fragments being shared among the commands. If there's little or no sharing, I think it's OK to leave as is. 90% of the commands required only the mutation string and that `pclient` call, so quite simple to change/maintain I think.

File: cylc/flow/scripts/cylc_ping.py (hunk 4 above)
Comment (oliver-sanders) [reply cylc/flow/scripts/cylc_ping.py:64]: > if the data-store had more than one workflow or the resolvers used scan to send the query off to other workflows

???

Comment (oliver-sanders) [reply cylc/flow/scripts/cylc_ping.py:64]: If the UIS is present we can get a workflow listing from there.

If the UIS is not present, how would this work? Surely we would need to do a FS scan to discover workflows?

File: cylc/flow/scripts/cylc_show.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/scripts/cylc_show.py:267]: I think ansimarkup has been configured to auto-close tags on each line.

Having said that, this may well be a mistake!

File: cylc/flow/scripts/cylc_ping.py (hunk 4 above)
Comment (dwsutherland) [reply cylc/flow/scripts/cylc_ping.py:64]: > If the UIS is present we can get a workflow listing from there.

Yes, this is already in place. The UIS mutation resolvers are different from the WFS, they parse the `ids`/`workflows` arguments to figure out which workflow(s) to send it to.

> If the UIS is not present, how would this work? Surely we would need to do a FS scan to discover workflows?

Correct, we would have to use resolvers like those at the UIS to parse the arguments against a list of workflow IDs.. However, we wouldn't want duplicate parsing at both UIS and WFS (so that might have to be an arg too!)

Comment (dwsutherland) [reply cylc/flow/scripts/cylc_ping.py:64]: You can actually test it out in the WUI GraphiQL, just use a glob for the workflow IDs i.e.:
```
mutation {
  stop(workflows: ["*"]) {
    result
  }
}
```
(off the top of my head)

**Comment by hjoliver** [comment]
(This is my next review priority)

**Review by hjoliver — COMMENTED** [review]
7/55 files viewed ... getting there :+1:

**Comment by hjoliver** [comment]
> > No change log entry required (invisible to users).
> 
> The only change I can think of that would effect users, is the CLI JSON format is different for a few commands (i.e. `cylc show` )..
> The other one; `cylc show --list-prerequisites` only shows prerequisites for live tasks, which may be rectified post n-window (TODO in code)..

:+1: not major problems.

File: cylc/flow/suite_status.py [hunk 8]
```diff
@@ -65,6 +65,9 @@ class StopMode(Enum):
     REQUEST_CLEAN = 'REQUEST(CLEAN)'
     """External shutdown request, will wait for active jobs to complete."""
 
+    REQUEST_KILL = 'REQUEST(KILL)'
+    """External shutdown request, will wait for active jobs to be killed."""
```
Comment (hjoliver) [inline cylc/flow/suite_status.py:69]: Not the fault of this PR (`git blame` fingers @oliver-sanders) but I don't think we doing this in other modules (are we?) and in fact last I heard Python doesn't have attribute docstrings - the PEP was rejected years ago.  So these strings should probably just be comments before the attributes?  (Or is this for Sphinx doc generation? I'm sure @oliver-sanders will enlighten me!)

Comment (dwsutherland) [reply cylc/flow/suite_status.py:69]: Yeah, it appears only the describe method is used to get different strings (?):
https://github.com/dwsutherland/cylc-flow/blob/e4c6e4cc9571893d6c7b9523d0cb1dbec9c383cc/cylc/flow/network/schema.py#L1297-L1308

File: tests/functional/graph-equivalence/00-oneline.t [hunk 9]
```diff
@@ -25,24 +25,21 @@ set_test_number 5
...
 TEST_NAME="${TEST_NAME_BASE}-validate"
-run_ok "${TEST_NAME}" cylc validate "${SUITE_NAME}"
+run_ok "${TEST_NAME}" cylc validate \
+    --set=TEST_OUTPUT_PATH="${PWD}" "${SUITE_NAME}"
 #-------------------------------------------------------------------------------
 TEST_NAME="${TEST_NAME_BASE}-run"
-suite_run_ok "${TEST_NAME}" \
-    cylc run --reference-test --debug --no-detach "${SUITE_NAME}"
+suite_run_ok "${TEST_NAME}" cylc run --reference-test --debug --no-detach \
+    --set=TEST_OUTPUT_PATH="${PWD}" "${SUITE_NAME}"
```
Comment (hjoliver) [inline tests/functional/graph-equivalence/00-oneline.t:33]: Ha-ha, I was wondering why the changes in this test and the associated `test1/flow.cylc`, then thought why in hell is this test passing on master... but I see you explained it:
> The other one; cylc show --list-prerequisites only shows prerequisites for live tasks, which may be rectified post n-window

Comment (hjoliver) [reply tests/functional/graph-equivalence/00-oneline.t:33]: (and same for the other graph-equivalence tests)

File: cylc/flow/data_store_mgr.py [hunk 10]
```diff
@@ -299,6 +306,7 @@ class DataStoreMgr:
         'min_point',
         'parents',
         'pool_points',
+        'publish_deltas',
```
Comment (hjoliver) [inline cylc/flow/data_store_mgr.py:309]: I should have picked this up previously, but we probably don't need `__slots__` in this class.  As I understand it, it's only a significant advantage (memory footprint) for classes that instantiate a large number of instances, such as task proxy and cycling-related objects in cylc-flow.

Comment (hjoliver) [reply cylc/flow/data_store_mgr.py:309]: Official docs on explain it as such, anyway:
https://docs.python.org/3/glossary.html#term-slots

> best reserved for rare cases where there are large numbers of instances in a memory-critical application.

File: cylc/flow/data_store_mgr.py [hunk 11]
```diff
@@ -1126,7 +1196,7 @@ def get_publish_deltas(self):
         result.append(
             (ALL_DELTAS.encode('utf-8'), all_deltas, 'SerializeToString')
         )
-        return result
+        return deepcopy(result)
```
Comment (hjoliver) [inline cylc/flow/data_store_mgr.py:1199]: Why this deepcopy?

File: cylc/flow/data_store_mgr.py (hunk 10 above)
Comment (dwsutherland) [reply cylc/flow/data_store_mgr.py:309]: Protobuf messages are class instances, does this count?

File: cylc/flow/data_store_mgr.py (hunk 11 above)
Comment (dwsutherland) [reply cylc/flow/data_store_mgr.py:1199]: So the data elements aren't modified between gathering them here and publishing them.

File: cylc/flow/data_store_mgr.py (hunk 10 above)
Comment (hjoliver) [reply cylc/flow/data_store_mgr.py:309]: It might help if the Protobuf message class itself had slots, but this is the `DataStoreMgr` class, which only gets instantiated once, right?

Comment (hjoliver) [reply cylc/flow/data_store_mgr.py:309]: i.e. wasted/spare memory in a small attributes dict doesn't matter much for a single instance, but it might for a very large number of instances.

Comment (dwsutherland) [reply cylc/flow/data_store_mgr.py:309]: Roger that, will remove slots (although it was helpful for the memory profiler to iterate through)

File: cylc/flow/data_store_mgr.py (hunk 11 above)
Comment (hjoliver) [reply cylc/flow/data_store_mgr.py:1199]: Fair enough.

**Review by hjoliver — APPROVED** [review]
Tests as working, passes all tests, code LGTM :+1: Question about use of slots is no biggie.
Great job @dwsutherland :bouquet: 

I still need to get a better understanding of some of the inner workings (and Protobuf unfortunately makes examining sh1t in a debugger more difficult, for me anyway :grimacing: )  but no need to block this (can I hit you up for a bit of a screen-sharing session Thursday or Friday @dwsutherland ?)

**Comment by hjoliver** [comment]
That's 2 approvals, but we should probably wait for at least a sanity check from @oliver-sanders on this one.

**Comment by dwsutherland** [comment]
> can I hit you up for a bit of a screen-sharing session Thursday or Friday @dwsutherland ?

For sure! I'll be out on Thursday afternoon, but Friday should be good if not tomorrow morning.

Thanks for the review!
(I'll get rid of slots in the n-window PR, since I'm modifying it there)

File: cylc/flow/data_store_mgr.py (hunk 11 above)
Comment (dwsutherland) [reply cylc/flow/data_store_mgr.py:1199]: I'll have to check if there's even a chance that they could be.. Might not be needed, probably safer, might depend on whether they need accessed aysnc to other things and how ZMQ publisher works..

File: cylc/flow/suite_status.py (hunk 8 above)
Comment (oliver-sanders) [reply cylc/flow/suite_status.py:69]: These strings do nothing in Python, however, Sphinx has some special parsing logic which picks them up which makes for a convenient way of documenting them.

File: cylc/flow/data_store_mgr.py (hunk 10 above)
Comment (oliver-sanders) [reply cylc/flow/data_store_mgr.py:309]: `__slots__` are unnecessary for classes like this that don't get initiated often, however, they are also harmless.

**Comment by oliver-sanders** [comment]
> That's 2 approvals, but we should probably wait for at least a sanity check from @oliver-sanders on this one.

I'm most the way there, haven't found anything significant, looks good.

This functionality is so heavily tested by the functional tests that I'm pretty happy it works.

File: tests/functional/graph-equivalence/02-splitline.t [hunk 12]
```diff
@@ -26,25 +26,22 @@ set_test_number 5
...
 #-------------------------------------------------------------------------------
 TEST_NAME="${TEST_NAME_BASE}-check-a"
-cylc run "${SUITE_NAME}" --hold 1>'out' 2>&1
-poll_grep_suite_log 'Holding all waiting or queued tasks now'
-cylc show "${SUITE_NAME}" 'a.1' | sed -n "/prerequisites/,/outputs/p" > 'a-prereqs'
 cmp_ok "${TEST_SOURCE_DIR}/splitline_refs/a-ref" 'a-prereqs'
 #-------------------------------------------------------------------------------
 TEST_NAME="${TEST_NAME_BASE}-check-b"
-cylc show "${SUITE_NAME}" 'b.1' | sed -n "/prerequisites/,/outputs/p" > 'b-prereqs'
+poll_grep_suite_log 'INFO - DONE'
```
[the remark this answers is not in the document]
Comment (dwsutherland) [reply tests/functional/graph-equivalence/02-splitline.t:40]: Hmm.. But the suite task runs the `cylc show` now for `b-prereqs`, so at minimum I'd need to poll the task for completion.

File: cylc/flow/data_store_mgr.py (hunk 10 above)
Comment (hjoliver) [reply cylc/flow/data_store_mgr.py:309]: Yeah no biggie, but still, I don't think we should use __slots__ unnecessarily. There is at least a small maintenance cost, and it makes the class slightly harder to use (in experimental mode, at least).

File: cylc/flow/data_store_mgr.py (hunk 11 above)
Comment (hjoliver) [reply cylc/flow/data_store_mgr.py:1199]: (That's one of the things I'd like to understand better).

File: cylc/flow/suite_status.py (hunk 8 above)
Comment (hjoliver) [reply cylc/flow/suite_status.py:69]: Should we have an issue up to extend this throughout the code then?

File: tests/functional/graph-equivalence/02-splitline.t (hunk 12 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply tests/functional/graph-equivalence/02-splitline.t:40]: You used `cylc run --no-detach` so the suite will run to completion before any of the `cmp_ok` tests run.

**Review by oliver-sanders — APPROVED** [review]
Finished code review, looks good, tested as working.

File: tests/functional/graph-equivalence/02-splitline.t (hunk 12 above)
[the remark this answers is not in the document]
Comment (dwsutherland) [reply tests/functional/graph-equivalence/02-splitline.t]: removed
